### PR TITLE
feat: add support for multimodal in Vertex

### PR DESCRIPTION
**Pull request by ArthurGoupil**
Following https://github.com/huggingface/chat-ui/pull/1021#issuecomment-2220018734

See https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/image-understanding#supported_models

### CODE REVIEW

**Comment by adhishthite** [comment]
@ArthurGoupil This is awesome. Will the same multimodal principles apply for OpenAI endpoints too? I am looking to develop multimodal capability for OpenAI endpoints

**Comment by ArthurGoupil** [comment]
Hi @adhishthite, I think you can reuse the generic parts of the implementation in https://github.com/huggingface/chat-ui/pull/1021, but the final data structure (with `inlineData`) is specific to Vertex, you should find the good one in OpenAI doc

**Review by nsarrazin — COMMENTED** [review]
Hi, thanks for the contribution, looks great! 

Would it be possible to also update the docs [here](https://github.com/huggingface/chat-ui/blob/main/docs/source/configuration/models/providers/google.md) to show how we can use multimodal models with vertex ?

Other than that LGTM!

File: README.md [hunk 1]
```diff
@@ -762,21 +762,33 @@ MODELS=`[
...
           "safetyThreshold": "BLOCK_MEDIUM_AND_ABOVE",
           "apiEndpoint": "", // alternative api endpoint url,
-          // Optional
           "tools": [{
             "googleSearchRetrieval": {
               "disableAttribution": true
             }
-          }]
+          }],
+          "multimodal": "{
```
Comment (pocman) [inline README.md:783]:
```suggestion
          "multimodal": {
```

File: README.md [hunk 2]
```diff
@@ -762,21 +762,33 @@ MODELS=`[
...
+            "image": {
+              "supportedMimeTypes": ["image/png", "image/jpeg", "image/webp"],
+              "preferredMimeType": "image/png",
+              "maxSizeInMB": 5,
+              "maxWidth": 2000,
+              "maxHeight": number;
+            }
+          }"
```
Comment (pocman) [inline README.md:791]:
```suggestion
          }
```

File: README.md [hunk 3]
```diff
@@ -762,21 +762,33 @@ MODELS=`[
...
             }
-          }]
+          }],
+          "multimodal": "{
+            "image": {
+              "supportedMimeTypes": ["image/png", "image/jpeg", "image/webp"],
+              "preferredMimeType": "image/png",
+              "maxSizeInMB": 5,
+              "maxWidth": 2000,
+              "maxHeight": number;
```
Comment (pocman) [inline README.md:789]:
```suggestion
              "maxHeight": 1000;
```

File: README.md [hunk 4]
```diff
@@ -762,21 +762,33 @@ MODELS=`[
...
+          "model": {
+            "id": "some-id",
+            "name": "gemini-1.5-pro-preview-0409"
+          },
 
           // Optional
           "safetyThreshold": "BLOCK_MEDIUM_AND_ABOVE",
           "apiEndpoint": "", // alternative api endpoint url,
-          // Optional
           "tools": [{
```
Comment (pocman) [inline README.md:778]: grounding is not supported with multimodal yet.
So tools or multimodal are mutualy exclusive

File: README.md [hunk 5]
```diff
@@ -762,21 +762,33 @@ MODELS=`[
     {
        "name": "gemini-1.5-pro",
        "displayName": "Vertex Gemini Pro 1.5",
+       "multimodal": true,
        "endpoints" : [{
           "type": "vertex",
           "project": "abc-xyz",
           "location": "europe-west3",
-          "model": "gemini-1.5-pro-preview-0409", // model-name
+          "model": {
```
Comment (pocman) [inline README.md:770]: This is not used, root name and id are used.

File: src/lib/server/endpoints/google/endpointVertex.ts [hunk 6]
```diff
@@ -43,7 +62,7 @@ export function endpointVertex(input: z.input<typeof endpointVertexParametersSch
 		const parameters = { ...model.parameters, ...generateSettings };
 
 		const generativeModel = vertex_ai.getGenerativeModel({
-			model: model.id ?? model.name,
+			model: model.name ?? model.id,
```
Comment (pocman) [inline src/lib/server/endpoints/google/endpointVertex.ts:65]: revert to model.id ?? model.name, until the nested model object is actually used.

**Review by pocman — COMMENTED** [review]
Tested locally, it's working, I'm able to upload an image and the model is using it to answer.

**Comment by nsarrazin** [comment]
Hi @ArthurGoupil, I'm not super familiar with the vertex endpoint but wdyt of the comments on the review from @pocman ? Feel free to approve or not and then I'll merge this :smile:

**Comment by ArthurGoupil** [comment]
Hi @nsarrazin, yeah i need to apply them once i have some time, probably in september :/

**Review by pocman — APPROVED** [review]

File: src/lib/server/endpoints/google/endpointVertex.ts [hunk 7]
```diff
@@ -73,7 +92,8 @@ export function endpointVertex(input: z.input<typeof endpointVertexParametersSch
 				stopSequences: parameters?.stop,
 				temperature: parameters?.temperature ?? 1,
 			},
-			tools,
+			// tools and multimodal are mutually exclusive
+			tools: !multimodal ? tools : undefined,
```
Comment (pocman) [inline src/lib/server/endpoints/google/endpointVertex.ts:96]: condition should not be on !multimodal but on files.length == 0

**Review by pocman — CHANGES_REQUESTED** [review]

**Comment by pocman** [comment]
@nsarrazin I believe this is ready to be merged. Can you have a look ?

**Pull request merged**